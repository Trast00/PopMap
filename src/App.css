* {
  padding: 0;
  margin: 0;
  box-sizing: border-box;
}

body {
  color: white;
  background-color: #ad3463;

  --color-primary-ligth: #ec4c8a;
  --color-primary: #db4680;
  --color-primary-dark: #cf4378;
}

a {
  text-decoration: none;
}

.flex-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.row {
  display: flex;
}

.column {
  display: flex;
  flex-direction: column;
}

.max-width {
  width: 100%;
  min-height: 300px;
}

.no-styling {
  list-style: none;
  text-decoration: none;
  border: none;
  background-color: transparent;
}

.App {
  text-align: center;
}

header {
  padding: 12px;
  justify-content: space-between;
}
