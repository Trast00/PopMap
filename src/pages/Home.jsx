import React from 'react'
import Header from '../components/Header'

const Home = () => {
  return (
    <div>
      <Header name="Home" canBack={false} />
      Page Home
    </div>
  )
}

export default Home