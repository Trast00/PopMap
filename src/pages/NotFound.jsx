import React from 'react';

const NotFound = () => (
  <div className="max-width flex-center column">
    <h2>404: Not Found</h2>
    <p>The page you re looking for does not exist</p>
    <p>Please go back to the home page</p>
  </div>
);

export default NotFound;
