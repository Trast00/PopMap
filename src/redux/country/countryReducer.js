import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import worldSvg from '../../assets/world.svg';

export const listContinents = [
  {
    name: "World",
    population: 0,
    flags: {
      svg: worldSvg,
    },
  },
  {
    name: "Africa",
    population: 0,
    flags: {
      svg: "",
    },
  },
  {
    name: "Americas",
    population: 0,
    flags: {
      svg: "",
    },
  },
  {
    name: "Asia",
    population: 0,
    flags: {
      svg: "",
    },
  },
  {
    name: "Europe",
    population: 0,
    flags: {
      svg: "",
    },
  },
  {
    name: "Oceania",
    population: 0,
    flags: {
      svg: "",
    },
  },
  {
    name: "Rest of World",
    population: 0,
    flags: {
      svg: "",
    },
  },
]

export const fetchData = createAsyncThunk(
  'popmap/country/GET_DATA',
  async () => {
    const result = await fetch('https://restcountries.com/v2/all');
    const data = await result.json();
    
    data.forEach(country => {
      listContinents[0].population += country.population
      switch(country.region){
        case "Africa": {
          listContinents[1].population += country.population
          break;
        }
        case "Americas": {
          listContinents[2].population += country.population
          break;
        }
        case "Asia": {
          listContinents[3].population += country.population
          break;
        }
        case "Europe": {
          listContinents[4].population += country.population
          break;
        }
        case "Oceania": {
          listContinents[5].population += country.population
          break;
        }
        default: {
          listContinents[6].population += country.population
          break;
        }
      }
    });

    return data;
  },
);

const countryReducer = createSlice({
  name: 'popmap/country/',
  initialState: {
    indexCurrentContinents: -1
  },
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchData.fulfilled,
      (state, { payload }) => ({ ...state, listCountries: [...payload] }));
  },
});


export default countryReducer.reducer;
