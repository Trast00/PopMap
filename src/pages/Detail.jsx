import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import DetailCountry from '../components/DetailCountry/DetailCountry';
import Header from '../components/Header';
import NotFound from './NotFound';

function Detail() {
  const { topLevelDomain } = useParams();

  const listCountry = useSelector((state) => state.countryReducer.listCountries);
  let name = topLevelDomain;
  let detailContent = '';
  if (listCountry) {
    const country = listCountry.filter((country) => (
      country.numericCode.toString() === topLevelDomain
    ))[0];
    if (country) {
      console.log('country is not null')
      name = country.name;
      detailContent = (<DetailCountry country={country} />);
    } else {
      // 404
      name = 'Error';
      detailContent = (<NotFound />);
    }
  }
  return (
    <div>
      <Header name={name} canBack />
      {listCountry && (detailContent)}
    </div>
  );
}

export default Detail;
