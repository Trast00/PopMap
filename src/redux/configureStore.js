import { configureStore } from '@reduxjs/toolkit';
import countryReducer from './country/countryReducer';

export default configureStore({
  reducer: {
    countryReducer,
  },
});
