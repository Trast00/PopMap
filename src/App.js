import './App.css';
import { Routes, Route } from 'react-router-dom';
import Home from './pages/Home';
import Detail from './pages/Detail';

const App = () => (
  <div className="App">
    <Routes>
      <Route exact path="/" element={<Home />} />
      <Route path="/:topLevelDomain" element={<Detail />} />
    </Routes>
  </div>
);

export default App;
