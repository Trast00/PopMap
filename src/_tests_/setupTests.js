import '@testing-library/jest-dom';
import configureStore from 'redux-mock-store';
import fetchMock from 'fetch-mock';
import thunk from 'redux-thunk';
import createMockStore from 'redux-mock-store';

export const testListCountry = [
  {
    name: "test country 1",
    population: 1000,
    flags: {
      svg: "test image url",
    },
    numericCode: "test1"
  },
  {
    name: "test country 2",
    population: 2000,
    flags: {
      svg: "test image url",
    },
    numericCode: "test2"
  }
]

const middleware = [thunk]
export const smockStore = createMockStore(middleware)

export const fetchDataMock = () => {
  fetchMock.getOnce('https://restcountries.com/v2/all', {
    headers: { 'content-type': 'application/json' },
    body: testListCountry,
  })
}