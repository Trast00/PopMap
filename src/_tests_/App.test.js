import renderer, { act } from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, MemoryRouter } from 'react-router-dom';
import React from 'react';
import App from '../App';
import configureStore from '../redux/configureStore'
import { fetchDataMock, testListCountry } from './setupTests.js';
import { fetchData } from '../redux/country/countryReducer';
import Home from '../pages/Home';
import { render, screen } from '@testing-library/react';
import Detail from '../pages/Detail';
import NotFound from '../pages/NotFound';

const store = configureStore;
fetchDataMock()

describe('test render', () => {
  test('App should match snapshoot', () => {
    const tree = renderer.create(
      <React.StrictMode>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App />
          </MemoryRouter>
        </Provider>
      </React.StrictMode>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('Test App render correctly after a mocked request', async () => { 
    await act(async () => {
      store.dispatch(fetchData);
    });

    const tree = renderer.create(
      <React.StrictMode>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/test2']}>
            <App />
          </MemoryRouter>
        </Provider>
      </React.StrictMode>,
    );

    await act(async () => {
      store.dispatch(fetchData);
    });

    const listCountry = store.getState().countryReducer.listCountries
    expect(listCountry).toMatchObject(testListCountry)

    expect(tree).toMatchSnapshot();
  })
});

jest.mock('../pages/Home')
jest.mock('../pages/Detail')
jest.mock('../pages/NotFound')

describe('test routes render', () => {
  test('should render home', () => { 
    Home.mockImplementation(()=> (<h1>Home page rendered</h1>))
    render(
      <React.StrictMode>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/']}>
            <App />
          </MemoryRouter>
        </Provider>
      </React.StrictMode>,
    )

    const pageTitle = screen.getByRole('heading')
    expect(pageTitle.textContent).toBe('Home page rendered')
  })

  test('should render detail', () => { 
    Detail.mockImplementation(()=> (<h1>Detail page rendered</h1>))
    render(
      <React.StrictMode>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/test1']}>
            <App />
          </MemoryRouter>
        </Provider>
      </React.StrictMode>,
    )

    const pageTitle = screen.getByRole('heading')
    expect(pageTitle.textContent).toBe('Detail page rendered')
  })
  test('should render detail', () => {
    Detail.mockImplementation(()=> (<NotFound />))
    NotFound.mockImplementation(()=>(<h1>Not found 404 rendred</h1>))
    render(
      <React.StrictMode>
        <Provider store={store}>
          <MemoryRouter initialEntries={['/test1020202']}>
            <App />
          </MemoryRouter>
        </Provider>
      </React.StrictMode>,
    )

    const pageTitle = screen.getByRole('heading')
    expect(pageTitle.textContent).toBe('Not found 404 rendred')
  })
})