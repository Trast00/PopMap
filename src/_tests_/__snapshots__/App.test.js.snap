// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test render List Rocket should match snapshoot 1`] = `
<div
  className="App"
/>
`;

exports[`test render Test App render correctly after a mocked request 1`] = `
<div
  className="App"
/>
`;
